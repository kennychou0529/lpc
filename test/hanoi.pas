program hanoi;
var n,nn,i,j:integer;
   M:array[1..2,1..1024] of integer;
   S:array[1..3,1..1024] of integer;
procedure hanoi1;
var k1,k2,k3,w,l,i,j,k:integer;
begin
 k1:=nn+1; k2:=k1 div 2; k3:=k2 div 2; M[1,k2]:=1; M[2,k2]:=2;
 for w:=n downto 2 do begin
   l:=k2;
   while l<=nn do begin
     i:=M[1,l]; j:=M[2,l]; k:=6-i-j;
     M[1,l-k3]:=i; M[2,l-k3]:=k;
     M[1,l+k3]:=k; M[2,l+k3]:=j;
     l:=l+k1; end;
    k1:=k2; k2:=k3; k3:=k3 div 2;  end;
end;

procedure hanoi2;
label 1;
var i,j,ns,niv,k:integer;
begin
 i:=1; j:=2; ns:=0; niv:=1;
 repeat
  k:=6-i-j;
  while niv<n-1 do begin
   ns:=ns+1; S[1,ns]:=i; S[2,ns]:=j; j:=k;
   k:=6-i-j; niv:=niv+1; s[3,ns]:=niv;
   end;
  write('(',i,',',k,')','(',i,',',j,')','(',k,',',j,')');
  if ns=0 then goto 1;
  i:=S[1,ns]; j:=S[2,ns]; niv:=S[3,ns]; ns:=ns-1;
  write('(',i,',',j,')');
  i:=6-i-j;
 until false;
1: end;

procedure hanoi3(n,i,j:integer);
var k:integer;
begin
 if n=1 then write('(',i,',',j,')')
        else begin  k:=6-i-j;
         hanoi3(n-1,i,k); write('(',i,',',j,')'); hanoi3(n-1,k,j); end;
end;

begin 
 writeln('HANOI'); write('n=');readln(n);writeln(n);
 writeln('1.HANOI1(umplerea matricii) 2.HANOI2(construire arbore complet)');
 writeln('3.HANOI3(recursiv)  4.EXIT');
 nn:=1; for i:=1 to n do nn:=2*nn; nn:=nn-1;
repeat
 write('optiunea dvs:');readln(i);
 case i of
  1: begin hanoi1;
      for i:=1 to 2 do begin
        for j:=1 to nn do write(M[i,j]:4);
             writeln; end; end;
  2: begin hanoi2; writeln; end;
  3: begin hanoi3(n,1,2); writeln end;
  4:writeln('BYE-BYE!')
end;
until i=4;

end.


