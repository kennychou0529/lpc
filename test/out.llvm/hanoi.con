
compiling pass 1 (LPC)
-----------------------------------------------------
  warning on line 6: 'i' hiding existing declaration on line 2
  warning on line 6: 'j' hiding existing declaration on line 2
  warning on line 21: 'i' hiding existing declaration on line 2
  warning on line 21: 'j' hiding existing declaration on line 2
  warning on line 38: 'n' hiding existing declaration on line 2
  warning on line 38: 'i' hiding existing declaration on line 2
  warning on line 38: 'j' hiding existing declaration on line 2

0 error(s), 7 warning(s)

Exit code: 0

compiling pass 2 (target = llvm)
-----------------------------------------------------

Invoking LLC ...
Exit code: 0

Linking...
Exit code: 0

running the executable
-----------------------------------------------------

Exit code: 0