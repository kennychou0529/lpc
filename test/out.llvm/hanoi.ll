; ModuleID = 'hanoi.pas'
source_filename = "C:\Users\lemo\work\compilers\lpc\test\hanoi.pas"
target datalayout = "e-m:w-i64:64-f80:128-n8:16:32:64-S128"

; runtime functions
declare dso_local i8* @_OpenFile(i32)
declare dso_local i8* @_OpenTempFile(i8*)
declare dso_local void @_CloseFile(i8*)
declare dso_local void @_WriteBool(i8*, i32, i32, i1)
declare dso_local void @_WriteChar(i8*, i32, i32, i8)
declare dso_local void @_WriteInteger(i8*, i32, i32, i32)
declare dso_local void @_WriteReal(i8*, i32, i32, double)
declare dso_local void @_WriteString(i8*, i32, i32, i8*, i32)
declare dso_local void @_WriteLn(i8*)

; program arguments (command line mapping)
%struct._Filename = type { i8*, i8* }
@_FilenameMapEntries = internal global [3 x %struct._Filename]
    [
    %struct._Filename { i8* getelementptr inbounds ([7 x i8], [7 x i8]* @.str.1, i32 0, i32 0), i8* null },
    %struct._Filename { i8* getelementptr inbounds ([8 x i8], [8 x i8]* @.str.2, i32 0, i32 0), i8* null },
    %struct._Filename { i8* null, i8* null }
    ], align 16
@_FilenameMap = dso_local global %struct._Filename* getelementptr inbounds ([3 x %struct._Filename], [3 x %struct._Filename]* @_FilenameMapEntries, i32 0, i32 0), align 8
@_FilenameMapSize = dso_local constant i32 2, align 4

; line 0
;================================================================================
; scope: global scope (level : 0)

; types
%T_text = type i8*

; line 1
;================================================================================
; scope: program scope (level : 1)

; types
%T_array_3 = type [1024 x i32]
%T_array_2 = type [2 x %T_array_3]
%T_array_5 = type [1024 x i32]
%T_array_4 = type [3 x %T_array_5]

; program variables
@_input = dso_local global %T_text zeroinitializer
@_output = dso_local global %T_text zeroinitializer
@i = dso_local global i32 zeroinitializer
@j = dso_local global i32 zeroinitializer
@M = dso_local global %T_array_2 zeroinitializer
@n = dso_local global i32 zeroinitializer
@nn = dso_local global i32 zeroinitializer
@S = dso_local global %T_array_4 zeroinitializer

; procedure body
define void @P_()
{
    ; initialize file handles
    %t1 = call i8* @_OpenFile(i32 0)
    store i8* %t1, i8** @_input
    %t2 = call i8* @_OpenFile(i32 1)
    store i8* %t2, i8** @_output

    ; cleanup
    %t3 = load %T_text, %T_text* @_output
    call void @_CloseFile(i8* %t3)
    %t4 = load %T_text, %T_text* @_input
    call void @_CloseFile(i8* %t4)

    ; epilogue
    ret void
}


; line 5
;================================================================================
; scope: hanoi1 (level : 2)

; activation record
%Frame_hanoi1 = type
{
    ; variables
    i32,    ; 0: i
    i32,    ; 1: j
    i32,    ; 2: k
    i32,    ; 3: k1
    i32,    ; 4: k2
    i32,    ; 5: k3
    i32,    ; 6: l
    i32,    ; 7: w

    ; dummy
    i8*
};

; procedure body
define void @P_hanoi1()
{
    ; allocate frame
    %.frame = alloca %Frame_hanoi1, align 8

    ; epilogue
    ret void
}


; line 19
;================================================================================
; scope: hanoi2 (level : 2)

; activation record
%Frame_hanoi2 = type
{
    ; variables
    i32,    ; 0: i
    i32,    ; 1: j
    i32,    ; 2: k
    i32,    ; 3: niv
    i32,    ; 4: ns

    ; dummy
    i8*
};

; procedure body
define void @P_hanoi2()
{
    ; allocate frame
    %.frame = alloca %Frame_hanoi2, align 8

    ; epilogue
    ret void
}


; line 38
;================================================================================
; scope: hanoi3 (level : 2)

; activation record
%Frame_hanoi3 = type
{
    ; parameters
    i32,    ; 0: i
    i32,    ; 1: j
    i32,    ; 2: n

    ; variables
    i32,    ; 3: k

    ; dummy
    i8*
};

; procedure body
define void @P_hanoi3(i32 %n, i32 %i, i32 %j)
{
    ; allocate frame
    %.frame = alloca %Frame_hanoi3, align 8
    %t1 = getelementptr inbounds %Frame_hanoi3, %Frame_hanoi3* %.frame, i32 0, i32 0
    store i32 %i, i32* %t1
    %t2 = getelementptr inbounds %Frame_hanoi3, %Frame_hanoi3* %.frame, i32 0, i32 1
    store i32 %j, i32* %t2
    %t3 = getelementptr inbounds %Frame_hanoi3, %Frame_hanoi3* %.frame, i32 0, i32 2
    store i32 %n, i32* %t3

    ; epilogue
    ret void
}


;================================================================================
; string literals

@.str.1 = private unnamed_addr constant [7 x i8] c"_input\00", align 1
@.str.2 = private unnamed_addr constant [8 x i8] c"_output\00", align 1


;================================================================================
; metadata

!llvm.dbg.cu = !{!24}
!llvm.module.flags = !{!25, !26}
!llvm.ident = !{!23}

!0 = !DIFile(filename: "hanoi.pas", directory: "C:\Users\lemo\work\compilers\lpc\test")
!1 = !DIBasicType(name: "boolean", size: 8, encoding: DW_ATE_boolean)
!2 = !DIBasicType(name: "char", size: 8, encoding: DW_ATE_unsigned_char)
!3 = !DIBasicType(name: "integer", size: 32, encoding: DW_ATE_signed)
!4 = !DIDerivedType(tag: DW_TAG_pointer_type, file: !0, line: 0, baseType: !5, size: 64)
!5 = !DIBasicType(name: "void", size: 0)
!6 = !DIBasicType(name: "real", size: 64, encoding: DW_ATE_float)
!7 = !DIDerivedType(tag: DW_TAG_pointer_type, name: "text", file: !0, line: 0, baseType: null, size: 64)
!8 = !DICompositeType(tag: DW_TAG_array_type, file: !0, line: 3, baseType: !9, size: 65536, elements: !13)
!9 = !DICompositeType(tag: DW_TAG_array_type, file: !0, line: 3, baseType: !3, size: 32768, elements: !11)
!10 = !DISubrange(count: 1024)
!11 = !{!10}
!12 = !DISubrange(count: 2)
!13 = !{!12}
!14 = !DICompositeType(tag: DW_TAG_array_type, file: !0, line: 4, baseType: !15, size: 98304, elements: !19)
!15 = !DICompositeType(tag: DW_TAG_array_type, file: !0, line: 4, baseType: !3, size: 32768, elements: !17)
!16 = !DISubrange(count: 1024)
!17 = !{!16}
!18 = !DISubrange(count: 3)
!19 = !{!18}
!20 = !{!1,!2,!3,!4,!5,!6,!7,!8,!9,!14,!15}
!21 = !{}
!22 = !{}
!23 = !{!"LPC 1.1"}
!24 = distinct !DICompileUnit(language: DW_LANG_Pascal83, file: !0, producer: "LPC 1.1", isOptimized: true, runtimeVersion: 0, emissionKind: FullDebug, enums: !21, retainedTypes: !20, globals: !22, nameTableKind: None)
!25 = !{i32 2, !"CodeView", i32 1}
!26 = !{i32 2, !"Debug Info Version", i32 3}

