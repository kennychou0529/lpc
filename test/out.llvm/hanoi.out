HANOI
n=4
1.HANOI1(umplerea matricii) 2.HANOI2(construire arbore complet)
3.HANOI3(recursiv)  4.EXIT
optiunea dvs:3375180916909059503976995052851433751809-16645373838210596743629050595071-165795801173757481011217951717222367078 262
16909059503976995052851433751809-16645373838210596743629050595071-165795801173757481011217951717222367078 262   1
optiunea dvs:(1,3)(1,2)(3,2)(197377,771)(-198142,197377)(-198142,771)(197377,771)(50528513,197377)(-50725884,50528513)(-50725884,197377)(50528513,197377)
optiunea dvs:(1,3)(1,2)(3,2)(1,3)(2,1)(2,3)(1,3)(1,2)(3,2)(3,1)(2,1)(3,2)(1,3)(1,2)(3,2)
optiunea dvs:BYE-BYE!
